### DISCUSSION TITLE
Combining createRoot, promises, and exceptions produces unexpected results

### DISCUSSION THREAD

**Discussion by jfrere** [solidjs/solid#1487] [Q&A] [answered]
Hi, I've run into this issue with `createRoot` where it produces results that my intuition can't explain.  I'm not sure if this behaviour is directly a bug, or if I'm doing things very wrong, or a bit of both!  😅

## Reproduction

Full reproduction: https://stackblitz.com/edit/solidjs-templates-pywzzx?file=src/index.tsx

## Description

My ultimate aim is to write tests for an asynchronous SolidJS hook in a fairly imperative manner (i.e. `obj = useMyHook(); await obj.thing(); assert(obj.result === 3)` - standard ACT-type stuff).  I've already found ["Testing your SolidJS code with jest"](https://dev.to/lexlohr/testing-your-solidjs-code-2gfh), and I've tried to use those ideas loosely, but also create a fairly generic wrapper so that I can move boilerplate code out of my tests and make them easier to understand.

Currently, my test wrapper/helper looks like this:

```typescript
function createTestRoot(testFn) {
  return () =>
    // create a new promise that will be resolved or rejected based on
    // the results of running `testFn` inside a created root.
    new Promise((resolve, reject) => {
      createRoot((dispose) => {
        const [resource, setResource] = createSignal({ state: 'STARTED' });

        // run `testFn` as a promise, and update the `resource` signal accordingly
        Promise.resolve((async () => testFn())()) // force testFn to be async even if it isn't
          .catch((error) => setResource({ state: 'ERRORED', error })
          .then(() => setResource({ state: 'FINISHED' }));

        // listen to the `resource` signal, and resolve or reject the 
        createEffect(() => {
          const resourceVal = resource();
          switch (resourceVal.state) {
            case 'STARTED':
              return;
            case 'FINISHED': {
              dispose();
              resolve();
              return;
            }
            case 'ERRORED': {
              dispose();
              reject(resourceVal.error);
              return;
            }
          }
        });
      });
    });
}
```

I use it to write tests that look something like this:

```typescript

test(
  "returns the earliest found date and the latest found date from trigger executions",
  createTestRoot(async () => {
    const quartzStore = createQuartzStore({
      fetchJobList: apiCall([ /* ... dummy data ... */ ]),
    });

    await quartzStore.loadAllJobs();
    expect(quartzStore.maximumTimespan.earliest).toEqual(
      new Date(2022, 1, 1)
    );
    expect(quartzStore.maximumTimespan.latest).toEqual(
      endOfDay(new Date(2025, 11, 12))
    );
  })
);
```

I wasn't 100% sure if this was the right option, but it seemed to work fairly well, and I was able to write a bunch of tests using it.  If there are no errors, this seems to work exactly as expected.

However, then I had a test that threw an error during the execution, and the error was completely swallowed, and the test started timing out.  It took me a while to figure out exactly what the problem was, but the smallest reproduction I could make (see https://stackblitz.com/edit/solidjs-templates-pywzzx?file=src%2Findex.tsx), seemed to require the following things:

* The error must be thrown after an `await` expression.  In the reproduction example above, if I remove the `await Promise.resolve()` line in the test, the exception is handled as expected.
* The error must be thrown during a `setState` (or I assume `setSignal`) call.  In the reproduction example above, if I throw the exception anywhere else, then the exception is handled as expected.
* An error must be thrown in the first place.  If the `setState` function succeeds, then the `createTestRoot` function works exactly as expected.  This makes me think that there is a bug happening somewhere, as I would expect `createTestRoot` to work the same whether or not the `testFn` promise throws or not.

I assume this has something to do with the issue described in the docs under `getOwner` and `runWithOwner`, but from the documentation and the example given, I'm not entirely sure how to use them to solve this problem.

## Questions

1. Is my code correct?  I wrote it very naively, and I would be very surprised if it is correct - but it was at least working fine until I ran into this issue!  😉
2. Is there a better way of solving my overall aim?  My issue with the example in the blog post above was that it was very verbose, and the train of logic jumped backwards and forwards a lot.  What I want is to be able to execute functions, `await` the results, and check that they and the store have all been updated correctly.  Is there a built-in solution here, or is that just not a very "SolidJS"-thing to want?
3. Is there a way of making this less confusing in the future?  What surprised me was that the error was swallowed (and in this one specific way as well), but tests without errors were working as I had expected.
    * If this is a problem that I've brought on myself, is there a way of logging a warning when SolidJS realises that something's gone wrong?  E.g. "you can't run `setState` here, because I don't know what the current root is" or something.

Thanks so much for your help!

**Comment by ninevra** [comment, marked as answer]
@jfrere can you try updating `solid-js` to the latest version? I'm pretty sure this is an example of #1498, which was fixed in version 1.6.10.  (`setStore()` uses `batch()` internally, and throwing exceptions in `batch()` after `await` used to break things.)

Running
```sh
pnpm up solid-js
vite --force
```
in your stackblitz reproduction fixes the issue for me.

**Reply by jfrere** [reply]
Perfect, that's fixed things completely.  Thanks for letting me know, and for the fix!
